File: memory/memory.cpp
#include <iostream>
#include <memory>
#include <string>

// Function that accepts a const reference
void f_ref(const int32_t& x) {
    std::cout << "Received const reference to: " << x << std::endl;
}

// Function that takes ownership through move
void f_move(std::unique_ptr<int32_t> x) {
    std::cout << "Took ownership of: " << *x << std::endl;
}

int main() {
    // "Use const references" example
    int32_t a = 3;
    std::cout << "Example: Use const references" << std::endl;
    f_ref(a);
    std::cout << "After reference use, a = " << a << std::endl;

    // "Don't copy, move" example
    std::cout << "\nExample: Don't copy, move" << std::endl;
    auto b = std::make_unique<int32_t>(3);
    std::cout << "Before move, b points to value: " << *b << std::endl;
    f_move(std::move(b));
    std::cout << "After move, b is " << (b ? "still valid" : "null") << std::endl;

    // Another move example with a string
    std::cout << "\nMoving a string example:" << std::endl;
    auto str = std::make_unique<std::string>("hello");
    std::cout << "Before move, str = " << *str << std::endl;
    auto str2 = std::move(str);
    std::cout << "After move, str is " << (str ? "still valid" : "null") << std::endl;
    std::cout << "str2 now owns the data: " << *str2 << std::endl;

    return 0;
}

File: memory/memory.rs
// Implementation of memory examples from README.md

// Function that accepts a reference (const reference example)
fn f_ref(x: &i32) {
    println!("Received reference to: {}", x);
}

// Function that takes ownership (move example)
fn f_move(x: i32) {
    println!("Took ownership of: {}", x);
}

fn main() {
    // "Use const references" example
    let a: i32 = 3;
    println!("Example: Use const references");
    f_ref(&a);
    println!("After reference use, a = {}", a);

    // "Don't copy, move" example
    let a: i32 = 3;
    println!("Example: Don't copy, move");
    f_move(a);
    // In Rust, primitive types like i32 implement Copy,
    // so we can still use 'a' after passing it to f_move
    println!("In Rust, primitives implement Copy, so a = {}", a);

    // To demonstrate true move semantics, let's use a non-Copy type like String
    let s = String::from("hello");
    println!("Moving a String, s = {}", s);
    let s2 = s;  // s is moved to s2
    // println!("Can't use s anymore: {}", s); // This would cause a compile error
    println!("s2 now owns the data: {}", s2);
}
